Level 3

mign0n opened the pull request.

mign0n committed:
Updated ReadMe file.

mign0n committed:
Fixed some minor inaccuracies.

mign0n committed:
Implemented /cities command in the telegram chatbot.

mign0n committed:
Implemented /cities command in the telegram chatbot.

Roxe322 commented on lpmegabot.py:
```diff
@@ -67,6 +69,51 @@ def get_next_full_moon(update, context):
+        if user_city in remaining_cities:
+            remaining_cities.remove(user_city)
+            last_letter = get_letter(user_city)
+            bot_city = choice([city for city in remaining_cities if city[0] == last_letter.upper()])
```
Это на самом деле очень тяжелая операция, ты проходишься по всему своему огромному списку городов, чтобы найти город на нужную букву. Более оптимальным решением было бы использовать словарь, где ключи будут буквами, а значения - сеты городов на эту букву

Roxe322 commented on lpmegabot.py:
```diff
@@ -67,6 +69,51 @@ def get_next_full_moon(update, context):
+
+
+def cities_game(update, context):
+    if context.args:
```
Предлагаю изучить конструкцию Guard Block, это когда ты инвертируешь условие, чтобы не создавать лишнюю вложенность 
Вот пример:

```
def a(b):
    if b:
        c = b + 1
        return c


def guard_block(b):
    if not b:
        return

    c = b + 1
    return c
```

Roxe322 replied:
Это стоит везде где возможно применять

Roxe322 commented on lpmegabot.py:
```diff
@@ -67,6 +69,51 @@ def get_next_full_moon(update, context):
+        context.user_data['bot_city_prev'] = bot_city
+        context.user_data['remaining_cities'] = remaining_cities
+        if notice is None:
+            update.message.reply_text(f"{bot_city}, Ваш ход.")
```
у тебя внутри if вызов одной и той же функции. Может стоит сделть переменную например response, которую из ифа заполнять, а вызывать одинаково. Плюс тут удобно можно воспользоваться тернарным оператором, например

`response = notice if notice is not None else 'Ваш ход`

mign0n committed:
Optimized search for the desired city by starting letter:
- Redesigned the method for storing city sets. Now cities are stored in a dictionary where keys are letters and values are sets of cities starting with this letter;
- Redesigned `cities_game_logic` for working with new data structure.

mign0n committed:
The following improvements were made:
- fixed a bug: the bot didn't recognize city names from more than 1 word;
- added the bot's response to the `/cities` command without an argument;
- fixed output city names according to the rules of the Russian language;
- fixed some minor inaccuracies.

mign0n committed:
Implemented  command in chatbot

mign0n committed:
Updated requirements.

Roxe322 commented on lpmegabot.py:
```diff
@@ -67,6 +70,82 @@ def get_next_full_moon(update, context):
+        return
+    expr = ''.join(context.args)
+    try:
+        result = ne.evaluate(expr)
```
Я бы не стал эту библиотеку применять для таких простых вещей. Это обвязка над питонячьим eval() в какой-то степени, а eval использовать небезопасно, тк там становятся доступны твои переменные. Тут лучше как-то более явно ограничивать входные данные

Roxe322 commented on lpmegabot.py:
```diff
@@ -67,6 +70,82 @@ def get_next_full_moon(update, context):
+        return None
+    else:
+        word = word.lower()
+        if position == 'first':
```
Вместо строковых представлений лучше использовать какие-то Enum-ограниченные значения, например завести класс Position c полями Position.first = 1, Position.last = -1, это и повышает читаемость кода и минимизирует возможность ошибок

Roxe322 commented on lpmegabot.py:
```diff
@@ -67,6 +70,82 @@ def get_next_full_moon(update, context):
+        if position == 'first':
+            letter = word[0]
+        else:
+            letter = word[-1] if word[-1] not in 'йьы' else word[-2]
```
Тут кажется не хватает твердого знака, а еще в теории возможна ситуация, где город заканчивается на две (или сколько угодно) неподдерживаемые буквы, что тоже стоит учитывать ;)

mign0n committed:
The following improvements were made:
- the logic of the city game is wrapped in a class;
- implemented recursive discarding of letters with which city names cannot begin;
- updated letters with which city names cannot begin;
- changed interface of function 'get_letter'.

mign0n committed:
Updated requirements.

mign0n committed:
Redisined calculator implementation.

Roxe322 commented on lpmegabot.py:
```diff
@@ -67,6 +69,112 @@ def get_next_full_moon(update, context):
+    for i in range(len(parts)):
+        parts[i] = parts[i].split('-')
+    for i in range(len(parts)):
+        parts[i] = first_priority_ops(parts[i])
```
А зачем тут два цикла одинаковых? Почему бы в одном не сделать? Да и вообще, я бы сделал это как-то в одном лист компрехеншене :) Попробуй сам, если что - напиши, я подскажу

mign0n committed:
Optimized calculator code.
